// @jupyterlab/cells/lib/model.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var coreutils_1 = require("@phosphor/coreutils");
var signaling_1 = require("@phosphor/signaling");
var codeeditor_1 = require("@jupyterlab/codeeditor");
var coreutils_2 = require("@jupyterlab/coreutils");
var outputarea_1 = require("@jupyterlab/outputarea");
/**
 * An implementation of the cell model.
 */
var CellModel = (function (_super) {
    __extends(CellModel, _super);
    /**
     * Construct a cell model from optional cell content.
     */
    function CellModel(options) {
        var _this = _super.call(this, { modelDB: options.modelDB }) || this;
        /**
         * A signal emitted when the state of the model changes.
         */
        _this.contentChanged = new signaling_1.Signal(_this);
        /**
         * A signal emitted when a model state changes.
         */
        _this.stateChanged = new signaling_1.Signal(_this);
        _this.id = options.id || coreutils_2.uuid();
        _this.value.changed.connect(_this.onGenericChange, _this);
        var cellType = _this.modelDB.createValue('type');
        cellType.set(_this.type);
        var observableMetadata = _this.modelDB.createMap('metadata');
        observableMetadata.changed.connect(_this.onGenericChange, _this);
        var cell = options.cell;
        var trusted = _this.modelDB.createValue('trusted');
        trusted.changed.connect(_this.onTrustedChanged, _this);
        if (!cell) {
            trusted.set(false);
            return _this;
        }
        trusted.set(!!cell.metadata['trusted']);
        delete cell.metadata['trusted'];
        if (Array.isArray(cell.source)) {
            _this.value.text = cell.source.join('\n');
        }
        else {
            _this.value.text = cell.source;
        }
        var metadata = coreutils_1.JSONExt.deepCopy(cell.metadata);
        if (_this.type !== 'raw') {
            delete metadata['format'];
        }
        if (_this.type !== 'code') {
            delete metadata['collapsed'];
            delete metadata['scrolled'];
        }
        for (var key in metadata) {
            observableMetadata.set(key, metadata[key]);
        }
        return _this;
    }
    Object.defineProperty(CellModel.prototype, "metadata", {
        /**
         * The metadata associated with the cell.
         */
        get: function () {
            return this.modelDB.get('metadata');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CellModel.prototype, "trusted", {
        /**
         * Get the trusted state of the model.
         */
        get: function () {
            return this.modelDB.getValue('trusted');
        },
        /**
         * Set the trusted state of the model.
         */
        set: function (newValue) {
            var oldValue = this.trusted;
            if (oldValue === newValue) {
                return;
            }
            this.modelDB.setValue('trusted', newValue);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Serialize the model to JSON.
     */
    CellModel.prototype.toJSON = function () {
        var metadata = Object.create(null);
        for (var _i = 0, _a = this.metadata.keys(); _i < _a.length; _i++) {
            var key = _a[_i];
            var value = JSON.parse(JSON.stringify(this.metadata.get(key)));
            metadata[key] = value;
        }
        if (this.trusted) {
            metadata['trusted'] = true;
        }
        return {
            cell_type: this.type,
            source: this.value.text,
            metadata: metadata,
        };
    };
    /**
     * Handle a change to the trusted state.
     *
     * The default implementation is a no-op.
     */
    CellModel.prototype.onTrustedChanged = function (trusted, args) { };
    /**
     * Handle a change to the observable value.
     */
    CellModel.prototype.onGenericChange = function () {
        this.contentChanged.emit(void 0);
    };
    return CellModel;
}(codeeditor_1.CodeEditor.Model));
exports.CellModel = CellModel;
/**
 * An implementation of a raw cell model.
 */
var RawCellModel = (function (_super) {
    __extends(RawCellModel, _super);
    function RawCellModel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(RawCellModel.prototype, "type", {
        /**
         * The type of the cell.
         */
        get: function () {
            return 'raw';
        },
        enumerable: true,
        configurable: true
    });
    return RawCellModel;
}(CellModel));
exports.RawCellModel = RawCellModel;
/**
 * An implementation of a markdown cell model.
 */
var MarkdownCellModel = (function (_super) {
    __extends(MarkdownCellModel, _super);
    /**
     * Construct a markdown cell model from optional cell content.
     */
    function MarkdownCellModel(options) {
        var _this = _super.call(this, options) || this;
        // Use the Github-flavored markdown mode.
        _this.mimeType = 'text/x-ipythongfm';
        return _this;
    }
    Object.defineProperty(MarkdownCellModel.prototype, "type", {
        /**
         * The type of the cell.
         */
        get: function () {
            return 'markdown';
        },
        enumerable: true,
        configurable: true
    });
    return MarkdownCellModel;
}(CellModel));
exports.MarkdownCellModel = MarkdownCellModel;
/**
 * An implementation of a code cell Model.
 */
var CodeCellModel = (function (_super) {
    __extends(CodeCellModel, _super);
    /**
     * Construct a new code cell with optional original cell content.
     */
    function CodeCellModel(options) {
        var _this = _super.call(this, options) || this;
        _this._outputs = null;
        var factory = (options.contentFactory ||
            CodeCellModel.defaultContentFactory);
        var trusted = _this.trusted;
        var cell = options.cell;
        var outputs = [];
        var executionCount = _this.modelDB.createValue('executionCount');
        if (!executionCount.get()) {
            if (cell && cell.cell_type === 'code') {
                executionCount.set(cell.execution_count || null);
                outputs = cell.outputs;
            }
            else {
                executionCount.set(null);
            }
        }
        executionCount.changed.connect(_this._onExecutionCountChanged, _this);
        _this._outputs = factory.createOutputArea({
            trusted: trusted,
            values: outputs,
            modelDB: _this.modelDB
        });
        _this._outputs.stateChanged.connect(_this.onGenericChange, _this);
        return _this;
    }
    Object.defineProperty(CodeCellModel.prototype, "type", {
        /**
         * The type of the cell.
         */
        get: function () {
            return 'code';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeCellModel.prototype, "executionCount", {
        /**
         * The execution count of the cell.
         */
        get: function () {
            return this.modelDB.getValue('executionCount');
        },
        set: function (newValue) {
            var oldValue = this.executionCount;
            if (newValue === oldValue) {
                return;
            }
            this.modelDB.setValue('executionCount', newValue || null);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeCellModel.prototype, "outputs", {
        /**
         * The cell outputs.
         */
        get: function () {
            return this._outputs;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources held by the model.
     */
    CodeCellModel.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        this._outputs.dispose();
        this._outputs = null;
        _super.prototype.dispose.call(this);
    };
    /**
     * Serialize the model to JSON.
     */
    CodeCellModel.prototype.toJSON = function () {
        var cell = _super.prototype.toJSON.call(this);
        cell.execution_count = this.executionCount || null;
        cell.outputs = this.outputs.toJSON();
        return cell;
    };
    /**
     * Handle a change to the trusted state.
     */
    CodeCellModel.prototype.onTrustedChanged = function (trusted, args) {
        if (this._outputs) {
            this._outputs.trusted = args.newValue;
        }
        this.stateChanged.emit({
            name: 'trusted',
            oldValue: args.oldValue,
            newValue: args.newValue
        });
    };
    /**
     * Handle a change to the execution count.
     */
    CodeCellModel.prototype._onExecutionCountChanged = function (count, args) {
        this.contentChanged.emit(void 0);
        this.stateChanged.emit({
            name: 'executionCount',
            oldValue: args.oldValue,
            newValue: args.newValue
        });
    };
    return CodeCellModel;
}(CellModel));
exports.CodeCellModel = CodeCellModel;
/**
 * The namespace for `CodeCellModel` statics.
 */
(function (CodeCellModel) {
    /**
     * The default implementation of an `IContentFactory`.
     */
    var ContentFactory = (function () {
        function ContentFactory() {
        }
        /**
         * Create an output area.
         */
        ContentFactory.prototype.createOutputArea = function (options) {
            return new outputarea_1.OutputAreaModel(options);
        };
        return ContentFactory;
    }());
    CodeCellModel.ContentFactory = ContentFactory;
    /**
     * The shared `ConetntFactory` instance.
     */
    CodeCellModel.defaultContentFactory = new ContentFactory();
})(CodeCellModel = exports.CodeCellModel || (exports.CodeCellModel = {}));
exports.CodeCellModel = CodeCellModel;
